docs: Cursor AI coding guidelines and an automated code quality

rvelaz opened the pull request:
### Context

As the MetaMask extension codebase grows and more developers (including AI coding agents) contribute, maintaining consistent code quality and adherence to coding standards becomes increasingly challenging. Without clear, accessible guidelines and automated enforcement, violations of best practices can slip through reviews, leading to technical debt and inconsistent code patterns.

### Problem

Currently, there's no centralized, easily accessible documentation for AI coding agents working on the MetaMask extension. This leads to:
- Inconsistent code patterns across contributions
- Repeated violations of coding standards (e.g., test naming conventions, component structure)
- Longer review cycles as reviewers catch common issues
- Difficulty onboarding new developers and AI agents to project standards
- Lack of automated code quality checks before PR creation

### Solution

This PR adds comprehensive AI coding agent guidelines and an automated code quality enforcement system:

**1. Cursor Rules Documentation** (`.cursor/rules/`):
- **Coding Guidelines** (`coding-guidelines/RULE.md`): TypeScript standards, React patterns, component structure, naming conventions, and code style
- **Controller Guidelines** (`controller-guidelines/RULE.md`): BaseController implementation patterns, state metadata requirements, selector patterns, and lifecycle management
- **Unit Testing Guidelines** (`unit-testing-guidelines/RULE.md`): Jest testing patterns, test structure, naming conventions, AAA pattern, async testing, and mocking strategies
- **E2E Testing Guidelines** (`e2e-testing-guidelines/RULE.md`): Playwright patterns, page object models, fixtures, wait strategies, and anti-patterns
- **Performance Guidelines**:
  - `front-end-performance-rendering/RULE.md`: Keys, memoization, virtualization, lazy loading
  - `front-end-performance-hooks-effects/RULE.md`: useEffect best practices, dependency arrays, cleanup
  - `front-end-performance-react-compiler/RULE.md`: React Compiler capabilities and when to manually optimize
  - `front-end-performance-state-management/RULE.md`: Redux optimization, selectors, normalization
- **Pull Request Guidelines** (`pull-request-guidelines/RULE.md`): PR creation, review best practices, and collaboration patterns

**2. CODEBOT Command** (`.cursor/commands/CODEBOT.md`):
- Automated analysis of changed files vs main branch
- Categorizes violations by severity (Critical, High, Medium)
- Provides actionable fix suggestions with line numbers
- PR readiness assessment with quality metrics
- Supports deep analysis mode for exhaustive checks
- Generates PR descriptions following the template

**3. Agent Instructions** (`AGENTS.md`):
- Comprehensive setup guide for AI agents
- Common workflows and commands
- Troubleshooting guide
- Performance optimization checklist
- Decision trees for common scenarios

### Implementation Notes

- All guideline files use the `.cursor/rules/` directory structure with `RULE.md` naming convention for consistency
- The E2E testing guidelines were reorganized from `.mdc` to `RULE.md` format
- CODEBOT analyzes only changed lines (not entire files) to focus on new code
- Guidelines reference official MetaMask documentation and contributor docs
- All examples follow MetaMask coding standards and patterns


### CODEBOT Usage

In Cursor chat:
```
/CODEBOT check my changes
/CODEBOT am I ready for PR?
/CODEBOT --deep # For exhaustive analysis
/CODEBOT --pr # Generate PR description only
```

# Example CODEBOT analysis:

https://github.com/MetaMask/metamask-extension/pull/38841 

# 🤖 CODEBOT: 38841 → ⚠️ NEEDS REVIEW

📋 **SUMMARY:** 2 critical, 2 high, 2 medium issues across 4 files

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
## 🔴 CRITICAL (2) - Must fix before merge
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

### 1. `test/e2e/tests/shield/shield-plan-subscription.spec.ts:275-276` - Debug console.log statements in test code

```typescript
console.log('shieldDetailPage 111 +++++++++++++++++++++++');
console.log('shieldDetailPage 222 !!!!!+++++++++++++++++++++++');
```

**Fix:** Remove debug console.log statements before committing:

```typescript
// Delete these lines entirely - they are debug statements
// that should not be committed to the codebase
```

📚 Rule: `.cursor/rules/coding-guidelines/RULE.md` → "Code Style" section: "Remove console.logs before committing"
Also: `.cursor/rules/e2e-testing-guidelines/RULE.md` → "Code Quality & Structure" checklist

---

### 2. `app/scripts/metamask-controller.js:1713` - Console.log in production code

```typescript
.catch((err) => {
  console.log('Error detecting tokens', err);
});
```

**Fix:** Use proper error logging or remove if unnecessary:

```typescript
.catch((err) => {
  // Use appropriate error handling for production
  log.error('Error detecting tokens', err);
  // Or if the error should be silently caught, add a comment explaining why
});
```

📚 Rule: `.cursor/rules/coding-guidelines/RULE.md` → "Code Style" section (lines 615-625): "No console.logs"

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
## 🟡 HIGH (2) - Should fix before merge
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

### 1. `test/e2e/tests/tokens/token-balances-controller-api-fallback.spec.ts:249,330,421` - ⚠️ DEPRECATED: Uses `driver.delay()` pattern

```typescript
// Line 249
await driver.delay(35000);

// Line 330
await driver.delay(35000);

// Line 421
await driver.delay(35000);
```

**Fix:** Use proper waiting strategies instead of hardcoded delays:

```typescript
// Instead of:
await driver.delay(35000);

// Use page object wait methods or waitForSelector:
await assetListPage.checkTokenAmountIsDisplayed('25 ETH');
// or
await driver.waitForSelector('[data-testid="asset-balance"]', { timeout: 35000 });
```

📚 Rule: `.cursor/rules/e2e-testing-guidelines/RULE.md` → "Proper Waiting and Assertions" section (lines 261-270)
Also: `.cursor/BUGBOT.md` → "Hard-coded delays" pattern

---

### 2. `test/e2e/tests/tokens/token-balances-controller-api-fallback.spec.ts` - Missing page object pattern for UI assertions

The test file uses direct driver calls and inline assertions instead of page object methods. For example, line 268:

```typescript
await assetListPage.checkTokenAmountIsDisplayed('25 ETH');
```

While `AssetListPage` is used, the test structure has some direct assertions and could benefit from more page object encapsulation for the verification logic.

**Recommendation:** Consider extracting the endpoint verification logic into a reusable helper or page object method.

📚 Rule: `.cursor/rules/e2e-testing-guidelines/RULE.md` → "Page Object Model (POM) Pattern" section (lines 113-164)

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
## 🔵 MEDIUM (2) - Consider fixing
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

### 1. `test/e2e/tests/tokens/token-balances-controller-api-fallback.spec.ts` - Large file with repeated mock setup code

The new E2E test file is 457 lines with significant duplication in mock setup (mockAccountApiV4Success, mockAccountApiV4Failure, mockRpcBalanceFallback share similar patterns). Consider extracting common mock patterns to a shared helper.

📚 Rule: `.cursor/rules/coding-guidelines/RULE.md` → "Avoid Repetitive Code (DRY Principle)" section

---

### 2. `test/e2e/helpers/shield/mocks.ts` - Very large file (437+ new lines)

The Shield mocks file has grown significantly with the addition of RPC mocks, gas API mocks, and transaction sentinel mocks. Consider breaking this into separate mock files by domain (e.g., `rpc-mocks.ts`, `gas-api-mocks.ts`).

📚 Rule: `.cursor/rules/coding-guidelines/RULE.md` → "File Organization" section

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
## 📝 PR DESCRIPTION (ready to copy)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

## **Description**

This PR adds support for Account API v4 with RPC fallback for token balance fetching. When the Account API is unavailable or returns an error, the TokenBalancesController gracefully falls back to direct RPC calls to fetch balances. Additionally, an `isOnboarded` callback has been added to both AccountTrackerController and TokenBalancesController initialization to prevent balance fetching before onboarding is complete.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/38841?quickstart=1)

## **Changelog**

CHANGELOG entry: Added token balance API fallback mechanism to ensure balances are displayed even when the Account API is unavailable

## **Related issues**

Fixes: #38841

## **Manual testing steps**

1. Build the extension with `yarn start`
2. Complete onboarding with a wallet that has tokens
3. Verify token balances are displayed on the home page
4. To test fallback: Use browser DevTools Network tab to block requests to `accounts.api.cx.metamask.io`
5. Refresh the extension and verify balances still appear (via RPC fallback)
6. On the token list, use the "Refresh" action from the control bar menu
7. Verify balances update correctly

## **Screenshots/Recordings**

N/A - No UI changes, backend functionality only

### **Before**
<!-- [screenshots/recordings] -->

### **After**
<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I've included tests if applicable
- [ ] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces a documented, automated quality workflow for AI and human contributors.
> 
> - **New** `CODEBOT` command in `.cursor/commands/CODEBOT.md` to analyze only changed lines vs `main`, classify violations (Critical/High/Medium) with evidence, assess PR readiness, and generate PR descriptions
> - **Adds** consolidated Cursor rules under `.cursor/rules/` (`coding-guidelines`, `controller-guidelines`, `e2e-testing-guidelines`, `front-end-performance-*`, `state-management`) with explicit checklists, examples, and selector/memoization guidance
> - **Standardizes** rule file structure (`RULE.md`, globs, metadata) for deterministic enforcement and references to MetaMask docs
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b1d3a3a937c919e50b7434069737be2cd425dd7d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

gauthierpetetin commented on AGENTS.md:
```diff
@@ -0,0 +1,1422 @@
+- [ ] LavaMoat policies updated (if dependencies changed)
+- [ ] Attributions updated (if dependencies changed)
+
+### Creating a PR
```
Here's I'd say the AI should create a PR description that exactly follows the [PR template](https://github.com/MetaMask/metamask-extension/blob/main/.github/pull-request-template.md).

gauthierpetetin commented on .cursor/commands/CODEBOT.md:
```diff
@@ -0,0 +1,614 @@
+description: Analyze code changes between current branch and main
+---
+
+# CODEBOT Command
```
This Cursor command is great 🚀

gauthierpetetin reviewed (commented):
Thanks a lot for putting this together, it's going to improve our use of AI 🚀

I've left two suggestions, but aside from that I think we should merge this.

One long-term thought: I'm wondering how we'll maintain these rules as contributor-docs evolve, since we'll then have kind of a duplicated source of truth for our contributor guidelines. Maybe it's not possible, but could the cursor rule just say "respect all guidelines from the contributor-docs repo"?

gauthierpetetin reviewed (dismissed):
LGTM

HowardBraham commented:
@rvelaz You have to do the CLA signature thing

rvelaz committed:
docs: add AI coding agent guidelines and documentation infrastructure

Add comprehensive documentation for AI agents including AGENTS.md, CODEBOT
command docs, and rule files covering coding standards, testing guidelines,
performance optimization, and PR workflows. Reorganize e2e-testing-guidelines
to RULE.md format.

rvelaz committed:
refactor: update guideline file paths in documentation

rvelaz commented:
I have read the CLA Document and I hereby sign the CLA

HowardBraham committed:
ran prettier --write

HowardBraham committed:
Make preventSinonAssertSyntax ignore MD, MDC, and JSON files

MajorLift commented on AGENTS.md:
```diff
@@ -0,0 +1,1635 @@
+### Prerequisites
+
+- **Node.js 24+** (use `nvm use` to auto-select from `.nvmrc`)
+- **Yarn 4.10.3+** (managed by Corepack, included with Node.js)
```
We should avoid fixed version numbers here

```suggestion
- **Node.js+** (use `nvm use` to auto-select development version specified in `.nvmrc`)
- **Yarn** (managed by Corepack, included with Node.js)
```

MajorLift commented on .cursor/commands/CODEBOT.md:
```diff
@@ -0,0 +1,1093 @@
+- [ ] Proper TypeScript types
+- [ ] No non-serializable values in state (no Promises, functions, Maps/Sets)
+- [ ] **No inline selector functions** in useSelector (extract to memoized selectors)
+- [ ] **No multiple useSelector calls** for same state slice (combine into one)
```
Shouldn't be universal rule as this is not always helpful for performance

```suggestion
```

MajorLift commented on .cursor/commands/CODEBOT.md:
```diff
@@ -0,0 +1,1093 @@
+- Each issue must be individually documented with full evidence
+- Developer must be able to verify the rule by opening the guideline file at specified lines
+- **Do NOT report pre-existing issues** in unchanged code
+
```
```suggestion
  - Attempt context compaction so that the report can be handed off to a new conversation, but only to the extent that the above requirements are still satisfied.

```

MajorLift reviewed (commented):
Some suggestions:

rvelaz committed:
Update .cursor/commands/CODEBOT.md

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

rvelaz committed:
Update AGENTS.md

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

rvelaz committed:
Update .cursor/commands/CODEBOT.md

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

rvelaz committed:
Update AGENTS.md

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

rvelaz committed:
Update .cursor/commands/CODEBOT.md

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

rvelaz committed:
Apply suggestions from code review

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

rvelaz committed:
Update AGENTS.md

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

MajorLift reviewed (approved):
LGTM! Very excited to see the output from this in actual prs.
